presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java

```diff
@@ -216,19 +259,182 @@ public OrcSelectiveRecordReader(
+        for (int i = 0; i < inputs.length; i++) {
+            int columnIndex = filterFunctionInputMapping.get(inputs[i]);
+            Object constantValue = constantValues[columnIndex];
+            blocks[i] = RunLengthEncodedBlock.create(columnTypes.get(columnIndex), constantValue == NULL_MARKER ? null : constantValue, 1);
+        }
+
+        initializeTmpErrors(1);
+        int positionCount = function.filter(new Page(blocks), new int[] {0}, 1, tmpErrors);
+
+        if (tmpErrors[0] != null) {
+            constantFilterError = tmpErrors[0];
+        }
+        return positionCount == 1;
+    }
+
+    private static boolean hasInputs(FilterFunction function)
+    {
+        return function.getInputChannels().length > 0;
+    }
+
+    private boolean allConstantInputs(FilterFunction function)
+    {
+        return Arrays.stream(function.getInputChannels())
+                .map(filterFunctionInputMapping::get)
+                .allMatch(columnIndex -> constantValues[columnIndex] != null);
+    }
+
+    private void reorderFiltersIfNeeded()
+    {
+        List<FilterFunctionWithStats> filters = Arrays.stream(filterFunctionsOrder)
+                .filter(Objects::nonNull)
+                .flatMap(functions -> functions.stream())
+                .sorted(Comparator.comparingDouble(function -> function.getStats().getElapsedNanonsPerDroppedPosition()))
+                .collect(toImmutableList());
+
+        assert filters.size() > 1;
+
+        Map<Integer, Integer> columnScore = new HashMap<>();
+        for (int i = 0; i < filters.size(); i++) {
+            int score = i;
+            Arrays.stream(filters.get(i).getFunction().getInputChannels())
+                    .map(filterFunctionInputMapping::get)
+                    // exclude columns with range filters
+                    .filter(columnIndex -> !columnsWithFilterScores.containsKey(columnIndex))
+                    // exclude constant columns
+                    .filter(columnIndex -> constantValues[columnIndex] == null)
+                    .forEach(columnIndex -> columnScore.compute(columnIndex, (k, v) -> v == null ? score : min(score, v)));
+        }
+
+        int[] newColumnOrder = columnScore.entrySet().stream()
+                .sorted(Comparator.comparing(Map.Entry::getValue))
+                .mapToInt(Map.Entry::getKey)
+                .toArray();
+
+        // Update streamReaderOrder,
+        // filterFunctionsOrder (aligned with streamReaderOrder),
+        // filterFunctionInputs (aligned with filterFunctionsOrder)
+        boolean sameOrder = true;
+        for (int i = 0; i < streamReaderOrder.length; i++) {
+            if (!columnsWithFilterScores.containsKey(streamReaderOrder[i])) {
+                for (int j = 0; j < newColumnOrder.length; j++) {
+                    if (streamReaderOrder[i] != newColumnOrder[j]) {
+                        sameOrder = false;
+                    }
+                    streamReaderOrder[i++] = newColumnOrder[j];
```

Comment:
is "i" getting double incremented in the loop.
Also why don't we inc both the arrays in a single loop
for(int i,j=0;condition;i++,j++)

Reply:
@bhhari `streamReaderOrder` contains 3 groups of columns, one after another: 

- columns with range filters;
- columns that provide inputs to filter functions;
- columns with no filter.

`newColumnOrder` contains columns from the second group in the "right" order.

The outer loop goes over streamReaderOrder until it reaches first column from the second group, then it loops over newColumnOrder to update columns in the second group, then exits.